File: Heap_Sort.cpp
//
//  main.cpp
//  Heap Sort
//  time complexity : O(nlogn)
//  Created by Feng Zhou on 10/3/16.
//  Copyright © 2016 Feng. All rights reserved.
//

#include <iostream>
#include <vector>

using namespace std;

void heap_insert(int input, vector<int> &output)
{
    output.push_back(input);
    auto index = output.size() - 1;
    while (index >= 1 && output[index] < output[(index - 1)/2]){
        swap(output[index],output[(index - 1)/2]);
        index = (index - 1) / 2;
    }
}
int heap_extract(vector<int> &source)
{
    int result = 0;
    if (source.size() > 0) {
        result = source[0];
    }
    // extract the root the put the last element to root
    swap(source[0],source[source.size() - 1]);
    source.pop_back();
    int index = 0;
    int minindex = 0;
    int minval = 0;
    while ((2*index +1) < source.size()){
        if (2*index + 2 == source.size()){
            //only one child for current node, so we regard it as the min value
            minval = source[2*index + 1];
            minindex = 2*index + 1;
        }
        else if (source[2*index + 1] < source[2*index + 2]) {
            minval = source[2*index + 1];
            minindex = 2*index + 1;
        }
        else{
            minval = source[2*index + 2];
            minindex = 2*index + 2;
        }
        if (source[index] > minval) {
            swap(source[index],source[minindex]);
            index = minindex;
        }
        else
            break;
    }
    return result;
}
vector<int> heap_sort(vector<int> v)
{
    vector<int> d;
    for (int i = 0 ; i < v.size(); i++) {
        heap_insert(v[i],d);
    }
    for (int i = 0 ; i < v.size(); i++)
    {
        v[i] = heap_extract(d);
    }
    return v;
}

int main(int argc, const char * argv[]) {
    // insert code here...
    vector<int> v = {3,4,51,12,31,5,4,6,7,13,100,80,51,23,42,12,523,2,0,1};
    v = heap_sort(v);
    for(int i = 0;i < v.size();i++)
        cout<<v[i]<<" ";
    cout << endl;
    return 0;
}

File: heap_sort.py
def heap_insert(x, l):
    l.append(x)
    n = len(l) - 1
    while n >= 1 and l[n] < l[int((n-1)/2)]:
        l[n], l[int((n-1)/2)] = l[int((n-1)/2)], l[n]
        n = int((n - 1)/2)
def heap_extract(l):
    # extract the root and put last to root
    l[0], l[len(l) - 1]  = l[len(l) - 1], l[0]
    res = l.pop(len(l) - 1)
    index, minindex, minval = 0, 0, 0
    while((2*index+1)<len(l)):
        if (2*index+2) == len(l):
            minval = l[2*index+1]
            minindex = 2*index+1
        elif l[2*index+1] < l[2*index+2]:
            minval = l[2*index+1]
            minindex = 2*index+1
        else :
            minval = l[2*index+2]
            minindex = 2*index+2
        if l[index] > minval:
            l[index], l[minindex] = l[minindex], l[index]
            index = minindex
        else:
            break
    return res
def heap_sort(l):
    r = list()
    for x in l:
        heap_insert(x, r)
    for i in range(len(l)):
        l[i] = heap_extract(r)

data = []
n = int(input('Enter how many elements you want: '))
for i in range(0, n):
    x = int(input('Enter the numbers into the array: '))
    data.append(x)
print('Original data is:', data)
heap_sort(data)
print('Sorted data is:', data)
